src/engine/gnc-pricedb.c

```diff
@@ -1345,68 +1356,147 @@ pricedb_remove_foreach_pricelist (gpointer key,
     g_list_foreach(node, (GFunc)check_one_price_date, data);
 
     LEAVE(" ");
 }
 
 static void
-pricedb_remove_foreach_currencies_hash (gpointer key,
-                                        gpointer val,
-                                        gpointer user_data)
+free_data (gpointer data)
 {
-    GHashTable *currencies_hash = (GHashTable *) val;
-
-    ENTER("key %p, value %p, data %p", key, val, user_data);
-    g_hash_table_foreach(currencies_hash,
-                         pricedb_remove_foreach_pricelist, user_data);
-    LEAVE(" ");
+    g_free (data);
 }
 
+static void
+gnc_pricedb_get_remove_dates (GHashTable* hash_dates, GDate period_begin, GDate period_end, PriceRemoveOptions leave)
+{
+    Recurrence *r;
+    GDate recurrance_date_old, recurrance_date_next;
+
+    GDateWeekday selected_day_of_week = G_DATE_FRIDAY;
+    GDate day_of_week_date;
+
+    day_of_week_date = period_begin;
+
+    r = g_new (Recurrence, 1);
+
+    DEBUG("Period Begin date is %d/%d/%d and End date is %d/%d/%d", g_date_get_day (&period_begin),
+           g_date_get_month (&period_begin), g_date_get_year (&period_begin), g_date_get_day (&period_end),
+           g_date_get_month (&period_end), g_date_get_year (&period_end));
+
+    g_date_set_day (&day_of_week_date, 1); // set date to 1st of month
+
+    while (g_date_get_weekday (&day_of_week_date) != selected_day_of_week)
+        g_date_subtract_days (&day_of_week_date, 1); // go back till we find last Friday of month
+
+    if (leave == PRICE_REMOVE_MONTHLY)
+        recurrenceSet (r, 1, PERIOD_LAST_WEEKDAY, &day_of_week_date, WEEKEND_ADJ_NONE); // Month set begin on last friday of month
+    else
+        recurrenceSet (r, 1, PERIOD_WEEK, &day_of_week_date, WEEKEND_ADJ_NONE); // Week set to begin on Friday of week
+
+    recurrance_date_next = recurrance_date_old = day_of_week_date;
+
+    while (g_date_compare (&period_end, &recurrance_date_next) > 0)
+    {
+        gchar *date_str;
+        
+        recurrenceNextInstance (r, &recurrance_date_old, &recurrance_date_next);
+        date_str = g_strdup_printf ("%d/%d/%d", g_date_get_day (&recurrance_date_old),
+                   g_date_get_month (&recurrance_date_old), g_date_get_year (&recurrance_date_old));
+
+        DEBUG("Recurrance date old string for hash table insert is %s", date_str);
+
+        g_hash_table_insert (hash_dates, date_str, "Keep"); // date_str freed by hashtable
+        recurrance_date_old = recurrance_date_next;
+    }
+    g_free (r);
+}
 
 gboolean
-gnc_pricedb_remove_old_prices(GNCPriceDB *db,
-                              Timespec cutoff,
-                              gboolean delete_user,
-                              gboolean delete_last)
+gnc_pricedb_remove_old_prices (GNCPriceDB *db, GList *comm_list,
+                              Timespec first, Timespec cutoff,
+                              PriceRemoveOptions source,
+                              PriceRemoveOptions leave)
```

Comment:
It's confusing that source and leave both use the same enum PriceRemoveOptions. This is also a potential source of nasty bugs if someone after you makes changes to this code and doesn't pay sufficient attention to which are the right combinations of options.